_posts/2023-06-13-http4s-authentication-part2.md

```diff
@@ -0,0 +1,1032 @@
+        barCodeData,
+        BarcodeFormat.QR_CODE,
+        width,
+        height
+      )
+      val outVal: FileOutputStream = new FileOutputStream(filePath)
+      MatrixToImageWriter.writeToStream(matrix, "png", outVal)
+    }
+}
+```
+
+The `createQRCode()` function creates the BarCode image. It takes as parameters, `barCodeData` which is our `URI`, a `filePath`, and `height` and `width` of our image, all of which we initialize to `barCode.png`, `400` and `400` respectively giving use an `IO[Unit]`.
+The `MultiFormatWriter` class has an `encode()` method that takes our arguments and creates a 2D matrix of bits.
+The `writeToStream()` method on the `MatrixToImageWriter` object takes the `matrix` (of type `BitMatrix`), `"png"` as the image format and a `FileOutputStream`, writes the image to a png file and saves it to the file path.
+
+Here's the full code:
+
+```scala
+package com.rockthejvm
+
+import com.rockthejvm.Generator.*
+import com.google.zxing.common.BitMatrix
+import com.google.zxing.MultiFormatWriter
+import com.google.zxing.BarcodeFormat
+import java.io.FileOutputStream
+import com.google.zxing.client.j2se.MatrixToImageWriter
+import cats.effect.IO
+
+object BarCodeService {
+  def getGoogleAuthenticatorBarCode(
+      secretKey: String,
+      account: String,
+      issuer: String,
+      generator: Generator
+  ): String =
+    generator match
+      case Hotp =>
+        s"otpauth://hotp/$issuer:$account?secret=$secretKey&issuer=$issuer&algorithm=SHA256&counter=0"
+      case Totp =>
+        s"otpauth://totp/$issuer:$account?secret=$secretKey&issuer=$issuer&algorithm=SHA256"
+
+  def createQRCode(
+      barCodeData: String,
+      filePath: String = "barCode.png",
+      height: Int = 400,
+      width: Int = 400
+  ): IO[Unit] =
+    IO {
+      val matrix: BitMatrix = new MultiFormatWriter().encode(
+        barCodeData,
+        BarcodeFormat.QR_CODE,
+        width,
+        height
+      )
+      val outVal: FileOutputStream = new FileOutputStream(filePath)
+      MatrixToImageWriter.writeToStream(matrix, "png", outVal)
+    }
+}
+```
+
+Let's create a `secret.env` file and store it in the root of our project. This will contain our the `secret` value that we will use for the next section. You can generate your own `secret` by using the code in the previous section:
+
+```bash
+export SECRET_KEY='VFI5NSVMHTC2SGYDFSJHVGOH4R4276XI'
+```
+
+Note, the env file should always be added to your .gitignore file for security purposes, and access to this file should also be limited by changing the file's permission.
+
+Create an `OtpService.scala` file that will contain our implementation for `Hotp` and `Totp`, and add the following code:
+
+```scala
+package com.rockthejvm
+
+import scala.util.*
+import com.rockthejvm.Generator
+
+class OtpService(generator: Generator, user: User) {
+  private val secret: Array[Byte] =
+    Properties.envOrElse("SECRET_KEY", "").getBytes()
```

Comment:
I found the problem. This secret key depends on the account and the issuer, it's not a static value. Check here: https://www.verifyr.com/en/otp/check#hotp